File: python/packages/agent_runtime/src/insightgraph_agent/orchestrator.py
from __future__ import annotations

import logging
from typing import Any

from pydantic import BaseModel, Field

from insightgraph_agent.analyst import Analyst
from insightgraph_agent.planner import Planner
from insightgraph_agent.retriever_agent import RetrieverAgent
from insightgraph_agent.session import SessionManager
from insightgraph_agent.verifier import Verifier

logger = logging.getLogger(__name__)


class AgentResponse(BaseModel):
    """The final response from the agent pipeline."""

    answer: str
    key_findings: list[str] = Field(default_factory=list)
    evidence: list[dict] = Field(default_factory=list)
    confidence: float = 0.0
    verified: bool = False
    question_type: str = ""
    steps_executed: int = 0


class Orchestrator:
    """Runs the Planner -> Retriever -> Analyst -> Verifier pipeline.

    Returns an evidence-backed answer with source citations.
    """

    def __init__(
        self,
        tools: Any,
        model: str = "gpt-4o-mini",
        api_key: str = "",
    ):
        self._planner = Planner(model=model, api_key=api_key)
        self._retriever = RetrieverAgent(tools)
        self._analyst = Analyst(model=model, api_key=api_key)
        self._verifier = Verifier(model=model, api_key=api_key)
        self._session_manager = SessionManager()

    @property
    def session_manager(self) -> SessionManager:
        return self._session_manager

    async def query(
        self,
        question: str,
        session_id: str | None = None,
    ) -> AgentResponse:
        """Process a question through the full agent pipeline.

        If session_id is provided, conversation context is included in planning
        and the session is updated with findings after the response.
        """
        # Get session context if available
        context = ""
        session = None
        if session_id:
            session = self._session_manager.get_session(session_id)
            if session:
                context = session.get_context_summary()

        # Step 1: Plan (with optional session context)
        logger.info("Planning for question: %s", question[:100])
        plan = await self._planner.plan(question, context=context)
        question_type = plan.get("question_type", "general")
        tool_plan = plan.get("tool_plan", [])

        # Step 2: Retrieve
        logger.info("Executing %d tool steps", len(tool_plan))
        retrieval_results = await self._retriever.execute_plan(tool_plan)

        # Step 3: Analyze
        logger.info("Analyzing results")
        analysis = await self._analyst.analyze(question, retrieval_results)

        # Step 4: Verify
        logger.info("Verifying analysis")
        verification = await self._verifier.verify(analysis, retrieval_results)

        # Build response
        confidence = verification.get("adjusted_confidence", analysis.get("confidence", 0.0))
        key_findings = analysis.get("key_findings", [])

        response = AgentResponse(
            answer=analysis.get("answer", "No answer available."),
            key_findings=key_findings,
            evidence=analysis.get("evidence_used", []),
            confidence=confidence,
            verified=verification.get("verified", False),
            question_type=question_type,
            steps_executed=len(tool_plan),
        )

        # Update session
        if session:
            entities_found = []
            for r in retrieval_results:
                result = r.get("result")
                if isinstance(result, list):
                    for item in result:
                        if isinstance(item, dict) and "name" in item:
                            entities_found.append(item["name"])
            session.add_turn(
                question=question,
                answer=response.answer,
                key_findings=key_findings,
                entities_found=entities_found,
            )

        return response

File: packages/agent-runtime/src/orchestrator.ts
import { Planner } from "./planner";
import { RetrieverAgent, ToolExecutor, StepResult } from "./retriever-agent";
import { Analyst } from "./analyst";
import { Verifier } from "./verifier";
import { SessionManager } from "./session";

/**
 * The final response from the agent pipeline.
 */
export interface AgentResponse {
  answer: string;
  keyFindings: string[];
  evidence: Array<Record<string, unknown>>;
  confidence: number;
  verified: boolean;
  questionType: string;
  stepsExecuted: number;
}

/**
 * Runs the Planner -> Retriever -> Analyst -> Verifier pipeline.
 * Returns an evidence-backed answer with source citations.
 */
export class Orchestrator {
  private planner: Planner;
  private retriever: RetrieverAgent;
  private analyst: Analyst;
  private verifier: Verifier;
  private _sessionManager: SessionManager;

  constructor(
    tools: ToolExecutor,
    model = "gpt-4o-mini",
    apiKey = "",
    baseUrl = "",
  ) {
    this.planner = new Planner(model, apiKey, baseUrl);
    this.retriever = new RetrieverAgent(tools);
    this.analyst = new Analyst(model, apiKey, baseUrl);
    this.verifier = new Verifier(model, apiKey, baseUrl);
    this._sessionManager = new SessionManager();
  }

  get sessionManager(): SessionManager {
    return this._sessionManager;
  }

  /**
   * Process a question through the full agent pipeline.
   *
   * If sessionId is provided, conversation context is included in planning
   * and the session is updated with findings after the response.
   */
  async query(question: string, sessionId?: string): Promise<AgentResponse> {
    // Get session context if available
    let context = "";
    const session = sessionId
      ? this._sessionManager.getSession(sessionId)
      : undefined;
    if (session) {
      context = session.getContextSummary();
    }

    // Step 1: Plan (with optional session context)
    console.info(`Planning for question: ${question.slice(0, 100)}`);
    const plan = await this.planner.plan(question, context);
    const questionType = plan.questionType;
    const toolPlan = plan.toolPlan;

    // Step 2: Retrieve
    console.info(`Executing ${toolPlan.length} tool steps`);
    const retrievalResults = await this.retriever.executePlan(toolPlan);
    const resultsAsRecords = retrievalResults as unknown as Array<Record<string, unknown>>;

    // Step 3: Analyze
    console.info("Analyzing results");
    const analysis = await this.analyst.analyze(question, resultsAsRecords);

    // Step 4: Verify
    //
    // NOTE: These could *not* safely be parallelized with the Analyst — the
    // Verifier's prompt inspects the Analyst's findings/citations and asserts
    // they're supported by the retrieval. Calling them concurrently with an
    // empty analysis object silently degrades verification quality to "always
    // verified", which is worse than a few hundred ms of latency.
    console.info("Verifying analysis");
    const verification = await this.verifier.verify(
      analysis as unknown as Record<string, unknown>,
      resultsAsRecords,
    );

    // Build response
    const confidence = verification.adjustedConfidence ?? analysis.confidence ?? 0.0;
    const keyFindings = analysis.keyFindings ?? [];

    const response: AgentResponse = {
      answer: analysis.answer ?? "No answer available.",
      keyFindings,
      evidence: analysis.evidenceUsed ?? [],
      confidence,
      verified: verification.verified ?? false,
      questionType,
      stepsExecuted: toolPlan.length,
    };

    // Update session
    if (session) {
      const entitiesFound: string[] = [];
      for (const r of retrievalResults) {
        const result = r.result;
        if (Array.isArray(result)) {
          for (const item of result) {
            if (item && typeof item === "object") {
              const rec = item as Record<string, unknown>;
              if (typeof rec.name === "string") {
                entitiesFound.push(rec.name);
              }
            }
          }
        }
      }
      session.addTurn(question, response.answer, keyFindings, entitiesFound);
    }

    return response;
  }
}
